File: csharp/simple-cipher/SimpleCipher.cs
using System;
using System.Linq;

public class SimpleCipher
{
    private enum Direction : int { 
        Left = -1, 
        Right = 1
    };

    const string alphabet = "abcdefghijklmnopqrstuvwxyz";

    private readonly string _key;

    public SimpleCipher() => _key = GenerateRandomKey();

    public SimpleCipher(string key) => _key = key;
    
    public string Key => _key;

    public string Encode(string plaintext) => Substitute(plaintext, Direction.Right);

    public string Decode(string ciphertext) => Substitute(ciphertext, Direction.Left);

    private string Substitute(string input, Direction direction) =>
        new string(input.Select((c, i) => SubstituteChar(c, i, direction)).ToArray());

    private char SubstituteChar(char c, int pos, Direction direction) {
        var offset = alphabet.IndexOf(_key[pos % _key.Length]) * (int) direction;

        // Add alphabet.Length to fix wrapping when going 'back'.
        // This is needed because in C# -1 mod 26 = -1, while mathematically it should be 25.
        return alphabet[(alphabet.IndexOf(c) + offset + alphabet.Length) % alphabet.Length];
    }

    private static string GenerateRandomKey() {
        const int keySize = 100;

        var rnd = new Random();
        var chars = from i in Enumerable.Repeat(alphabet.Length - 1, keySize)
                    select alphabet[rnd.Next(i)];

        return new string(chars.ToArray());
    }
}

File: kotlin/simple-cipher/src/main/kotlin/SimpleCipher.kt
private val keyRange = ('a'..'z').toList()
private fun randomKey() = List(100) { _ -> keyRange.random() }.joinToString("")

private enum class CipherMode {
    Encode, Decode
}

data class Cipher(val key: String) {

    init {
        require(key.isNotBlank())
        require(key.all { it in keyRange })
    }

    constructor() : this(randomKey())

    fun encode(s: String) = substitute(s, CipherMode.Encode)

    fun decode(s: String) = substitute(s, CipherMode.Decode)

    private fun substitute(s: String, mode: CipherMode) =
        s.mapIndexed { i, c -> substituteChar(c, i, mode) }.joinToString("")

    private fun substituteChar(char: Char, position: Int, mode: CipherMode): Char {
        val direction = when (mode) {
            CipherMode.Encode -> 1
            CipherMode.Decode -> -1
        }

        val offset = keyRange.indexOf(key[position % key.length]) * direction
        val index = (keyRange.indexOf(char) + offset).mod(keyRange.size)
        return keyRange[index]
    }
}
